### ISSUE TITLE
Arrays of type float96 or float128 do not display correctly (migrated from Trac #494)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#1101]
Original ticket http://projects.scipy.org/numpy/ticket/494
Reported 2007-04-05 by atmention:davidsocha, assigned to unknown.

Here is an example:

```
C:\Documents and Settings\socha>python
Python 2.4.3 - Enthought Edition 1.0.0 (#69, Aug  2 2006, 12:09:59) [MSC v.1310 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
>>> numpy.__version__
'1.0.2'
>>> from numpy import array
>>> f = array([1,2,3], dtype='float96')
>>> f
array([0.0, 0.0, -2.0], dtype=float96)             <== float96 prints incorrectly
>>> f.astype('float32')
array([ 1.,  2.,  3.], dtype=float32)              <== Yet data seems intact
>>> f.sum()
-2.0                                               <== prints incorrectly
>>> f.sum().astype('float32')
6.0                                                <== prints correctly if force to float32
>>>
```

We have the same symptoms with float128 (on a Mac with an Intel 2GHz Core Duo).
- David Socha & Daniel Terhorst, UrbanSim Project, http://www.urbansim.org

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2007-04-05

As a data point, I don't see this problem on my Linux installation with r3650.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2007-04-07

I don't see this either. Socha, could you reinstall numpy after removing the numpy folder in site-packages and the build subdirectory? I can't test the float128 version without installing a 64bit system.

**Comment by thouis** [comment]
Comment in Trac by atmention:davidsocha, 2007-04-08

Here's the code I tried (extracted from original submission):

```
import numpy
numpy.__version__
from numpy import array
f = array([1,2,3], dtype='float96')
f
f.astype('float32')
f.sum()
f.sum().astype('float32')
```

Here are the experiments I did:

(1) Removed numpy directory.  Reinstalled numpy from numpy-1.0.2.win32-py2.4.exe.  Same problem.

(2) Removed numpy and Numeric directories.  Reinstalled numpy.  Same problem.

(3) Removed numpy and numarray directories.  Reinstalled numpy.  Same problem.

Site-packages directory now has contents shown below.

(4) Removed numpy and reinstalled numpy from numpy-1.0.1.win32-py2.4.exe.  Same problem.

Note that I did not see this problem on our Linux computer.  However, I have seen this problem on both Windows computers that I have tried it on.

It may be that to duplicate this, you need to install Python Enthought Edition--Python 2.4.3 for Windows Enhanced Python Distribution : http://code.enthought.com/enthon/

Perhaps this is due to a conflict with another package?  Here is the contents of my site-packages directory after (3) above:

```
C:\Python24\Lib\site-packages>dir
 Volume in drive C has no label.
 Volume Serial Number is 20A5-3ACF

 Directory of C:\Python24\Lib\site-packages

04/08/2007  08:22 AM    <DIR>          .
04/08/2007  08:22 AM    <DIR>          ..
04/04/2007  11:42 AM    <DIR>          BTrees
07/26/2006  04:07 PM           147,968 cElementTree.pyd
04/04/2007  11:41 AM    <DIR>          ctypes
04/06/2007  02:21 PM    <DIR>          dateutil
04/03/2007  05:42 PM    <DIR>          dbfpy
04/04/2007  11:44 AM    <DIR>          docutils
12/21/2004  01:46 AM            10,012 dot_parser.py
06/30/2006  07:58 PM               216 easy-install.pth
04/04/2007  11:43 AM    <DIR>          elementtree
04/04/2007  11:46 AM    <DIR>          enthought
02/24/2005  11:42 AM             5,754 fpconst.py
04/04/2007  11:43 AM    <DIR>          gadfly
04/04/2007  11:43 AM    <DIR>          IPython
04/04/2007  11:44 AM    <DIR>          isapi
04/06/2007  02:21 PM    <DIR>          matplotlib
04/04/2007  11:40 AM    <DIR>          mayavi
04/03/2007  05:42 PM    <DIR>          MySQLdb
04/03/2007  05:42 PM    <DIR>          MySQL_python-1.2.2-py2.4.egg-info
04/08/2007  08:20 AM    <DIR>          numpy
04/04/2007  11:42 AM    <DIR>          persistent
04/04/2007  11:41 AM    <DIR>          PIL
08/02/2006  12:02 PM                 5 PIL.pth
04/04/2007  11:41 AM    <DIR>          PyCrust
12/21/2004  01:45 AM            33,616 pydot.py
07/07/2006  02:48 PM                32 pylab.py
02/06/2006  08:10 AM            99,402 pyparsing.py
04/04/2007  11:44 AM    <DIR>          pyreadline
04/04/2007  11:44 AM    <DIR>          pysqlite2
04/04/2007  11:44 AM    <DIR>          pysvn
10/31/2003  06:47 PM               141 pythoncom.py
04/04/2007  11:44 AM    <DIR>          pythonwin
04/06/2007  02:21 PM    <DIR>          pytz
07/25/2006  08:45 PM         2,045,202 PyWin32.chm
10/10/2004  05:15 PM                69 pywin32.pth
08/02/2006  12:05 PM                 5 pywin32.version.txt
04/04/2007  11:44 AM    <DIR>          pywin32_system32
05/07/2006  10:26 AM               811 readline.py
04/03/2007  02:36 PM               653 readline.pyc
03/31/2000  03:03 PM               119 README.txt
06/15/2003  10:23 PM             2,687 roman.py
04/04/2007  11:41 AM    <DIR>          Scientific
04/04/2007  01:18 PM    <DIR>          scipy
04/04/2007  11:44 AM    <DIR>          serial
04/04/2007  11:44 AM    <DIR>          setuptools-0.6a11dev-py2.4.egg
08/02/2006  12:07 PM                32 setuptools.pth
06/30/2006  07:58 PM             2,444 site.py
06/30/2006  08:00 PM             1,812 site.pyo
04/04/2007  11:43 AM    <DIR>          SOAPpy
04/04/2007  11:44 AM    <DIR>          tables
04/04/2007  11:44 AM    <DIR>          testoob
04/04/2007  11:42 AM    <DIR>          ThreadedAsync
04/04/2007  11:42 AM    <DIR>          transaction
04/04/2007  11:40 AM    <DIR>          vtk
04/20/2006  11:43 AM                10 vtk.pth
04/04/2007  11:40 AM    <DIR>          vtkPipeline
04/04/2007  11:40 AM    <DIR>          vtk_python
04/04/2007  11:44 AM    <DIR>          win32
04/04/2007  11:45 AM    <DIR>          win32com
04/03/2007  11:08 AM    <DIR>          win32comext
04/03/2007  11:06 AM    <DIR>          wx-2.6-msw-unicode-enthought
04/19/2006  01:18 PM                28 wx.pth
05/16/2005  06:36 PM            17,511 wxversion.py
04/04/2007  11:42 AM    <DIR>          ZConfig
04/04/2007  11:43 AM    <DIR>          zdaemon
04/04/2007  11:43 AM    <DIR>          ZEO
04/04/2007  11:43 AM    <DIR>          ZODB
04/04/2007  11:43 AM    <DIR>          zope
04/04/2007  11:43 AM    <DIR>          ZopeUndo
07/26/2006  04:08 PM            86,016 _ctypes.pyd
07/26/2006  04:08 PM            11,776 _ctypes_test.pyd
04/04/2007  12:57 PM         1,507,328 _mysql.pyd
04/04/2007  12:57 PM             2,306 _mysql_exceptions.py
04/04/2007  12:57 PM             3,680 _mysql_exceptions.pyc
04/04/2007  12:57 PM             3,680 _mysql_exceptions.pyo
04/04/2007  11:43 AM    <DIR>          _xmlplus
              28 File(s)      3,983,315 bytes
              49 Dir(s)  33,296,609,280 bytes free
```

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2007-04-08

I was suspicious of the windows compiler when I saw MSC v.1310 in the python startup string, but the fact that the problems also occurred on the Mac made that less likely. Are you running windows on the Mac hardware? Also, it looks like you are installing the precompiled numpy  releases, is that the case? Have you tried compiling from scratch?

Thanks for the extensive testing. I need to boot up a windows computer somewhere and see just how the MSC compiler handles long doubles. The compiler itself looks like a pretty old version.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2007-04-08

Looks like the print routine is stepping correctly through the array but treating the resulting values as doubles. Here is a hoked up test where every third value (2*96/64) should correspond to your numbers 1,2,3

```
In [16]: f = fromstring(array([1,0,2,0,3,0],float96).tostring(),dtype=double)

In [17]: f[::3]
Out[17]: array([-0., -0., -2.])
```

So it looks like a format string problem, although -0.0 has a sign that doesn't appear in your output. I find it odd that the formats should vary depending on the compiler. Hmmm...

**Comment by thouis** [comment]
Comment in Trac by atmention:davidsocha, 2007-04-12

The results are from three different platforms:
- 2 Dell desktops running Windows XP Pro
- one MacBook (Pro?) running OS X
- several Dell servers running a modified Linux Fedora core 5

The problems occur on the Windows and Mac boxes; not on the Linux boxes.

On Windows and Mac, we installed numpy from the precompiled numpy releases.

On our Linux boxes, we compiled numpy from scratch and do not have this problem.

We have not tried compiling from scratch on either platform.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2007-05-13

Microsoft compilers treat the long double type as a double, at least up to VC++8.0

"So the two types are distinguishable, though both types are currently mapped to the same double precision format; that is, sizeof(d) == sizeof(ld).  Much to my chagrin, VC++ still doesn’t have extended precision long doubles (yes, even in version 8.0!).  This has been a point of longstanding criticism of our product.  For years processors have provided the capability of extended precision, yet VC only provides support for double precision."

So this looks like a compiler problem. I think this ticket should be closed as invalid. Maybe another compiler should be used for the windows' platform? I don't know what is going on with OSX, I think it uses the gcc compiler, but I suspect there it is a compiler problem there also.

Heads up on another problem coming: MS Vista limits binaries compiled with "foreign" compilers to 32 meg of memory. It is said this will be fixed by SP1.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2007-05-13

And google also turns up the following:

"Applications that run on Mac OS X 10.3.9 must not try to use the long double type or any of the functions associated with it. Support for the 128-bit long double type was not introduced until Mac OS X 10.4. Similarly, full support for sinl, expl, and all other long double math functions defined by the C standard is available only in 10.4 and later."

I googled the words

OSX  compiler long double support

and

VC++ compiler long double support

to find this and the previous reference. So I am closing the ticket.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2007-05-13

I suppose there are a few things we could do.

1) on windows platforms make long doubles have the right size (float64), or return a not implemented error.

2) check the OS X version on the Mac and return not implemented if the version is < 10.4.
